src/metabase/query_processor/middleware/annotate.clj

```diff
@@ -183,81 +183,84 @@
-      (assoc :base_type (:base-type opts))
-
-      namespaced-options
-      (assoc :options namespaced-options)
-
-      (string? id-or-name)
-      (merge (or (some-> (some #(when (= (:name %) id-or-name) %) source-metadata)
-                         (dissoc :field_ref))
-                 {:name         id-or-name
-                  :display_name (humanization/name->human-readable-name id-or-name)}))
-
-      (integer? id-or-name)
-      (merge (let [{parent-id :parent_id, :as field} (dissoc (qp.store/field id-or-name) :database_type)]
-               (if-not parent-id
-                 field
-                 (let [parent (col-info-for-field-clause inner-query [:field parent-id nil])]
-                   (update field :name #(str (:name parent) \. %))))))
-
-      (:binning opts)
-      (assoc :binning_info (-> (:binning opts)
-                               (set/rename-keys {:strategy :binning-strategy})
-                               u/snake-keys))
-
-      (:temporal-unit opts)
-      (assoc :unit (:temporal-unit opts))
-
-      (or (:join-alias opts) (:alias join))
-      (assoc :source_alias (or (:join-alias opts) (:alias join)))
-
-      join
-      (update :display_name display-name-for-joined-field join)
-
-      ;; Join with fk-field-id => IMPLICIT JOIN
-      ;; Join w/o fk-field-id  => EXPLICIT JOIN
-      (:fk-field-id join)
-      (assoc :fk_field_id (:fk-field-id join))
-
-      ;; For IMPLICIT joins, remove `:join-alias` in the resulting Field ref -- it got added there during
-      ;; preprocessing by us, and wasn't there originally. Make sure the ref has `:source-field`.
-      (:fk-field-id join)
-      (update :field_ref mbql.u/update-field-options (fn [opts]
-                                                       (-> opts
-                                                           (dissoc :join-alias)
-                                                           (assoc :source-field (:fk-field-id join)))))
-
-      ;; If source Field (for an IMPLICIT join) is specified in either the field ref or matching join, make sure we
-      ;; return it as `fk_field_id`. (Not sure what situations it would actually be present in one but not the other
-      ;; -- but it's in the tests :confused:)
-      (or (:source-field opts)
-          (:fk-field-id join))
-      (assoc :fk_field_id (or (:source-field opts)
-                              (:fk-field-id join)))
-
-      ;; If the source query is from a saved question, remove the join alias as the caller should not be aware of joins
-      ;; happening inside the saved question. The `not join-is-at-current-level?` check is to ensure that we are not
-      ;; removing `:join-alias` from fields from the right side of the join.
-      (and source-card-id
-           (not join-is-at-current-level?))
-      (update :field_ref mbql.u/update-field-options dissoc :join-alias))))
+  (if (:canonical-field opts)
+    (col-info-for-field-clause* inner-query (:canonical-field opts))
```

Comment:
We put the field we replace in the join in fill-bins into the :canonical-field key on the field options.